src/H5Dio.c

```diff
@@ -137,363 +133,626 @@ H5D__read(H5D_t *dataset, hid_t mem_type_id, H5S_t *mem_space, H5S_t *file_space
+
+                /* Make second phase IO init call */
+                if ((dset_info[i].layout_ops.mdio_init)(&io_info, &(dset_info[i])) < 0)
+                    HGOTO_ERROR(H5E_DATASET, H5E_CANTINIT, FAIL, "can't populate array of selected pieces")
+
+                /* Reset metadata tagging */
+                H5AC_tag(prev_tag, NULL);
+            }
+
+        /* Invoke correct "high level" I/O routine */
+        if ((*io_info.md_io_ops.multi_read_md)(&io_info) < 0)
+            HGOTO_ERROR(H5E_DATASET, H5E_READERROR, FAIL, "can't read data")
+    } /* end if */
+    else {
+        haddr_t prev_tag = HADDR_UNDEF;
+
+        if (!H5D_LAYOUT_CB_PERFORM_IO(&io_info) && io_info.piece_count > 0) {
+            if (NULL == (io_info.mem_spaces = H5MM_malloc(io_info.piece_count * sizeof(H5S_t *))))
+                HGOTO_ERROR(H5E_RESOURCE, H5E_CANTALLOC, FAIL,
+                            "memory allocation failed for memory space list")
+            if (NULL == (io_info.file_spaces = H5MM_malloc(io_info.piece_count * sizeof(H5S_t *))))
+                HGOTO_ERROR(H5E_RESOURCE, H5E_CANTALLOC, FAIL, "memory allocation failed for file space list")
+            if (NULL == (io_info.addrs = H5MM_malloc(io_info.piece_count * sizeof(haddr_t))))
+                HGOTO_ERROR(H5E_RESOURCE, H5E_CANTALLOC, FAIL,
+                            "memory allocation failed for piece address list")
+            if (NULL == (io_info.element_sizes = H5MM_malloc(io_info.piece_count * sizeof(size_t))))
+                HGOTO_ERROR(H5E_RESOURCE, H5E_CANTALLOC, FAIL,
+                            "memory allocation failed for element size list")
+            if (NULL == (io_info.rbufs = H5MM_malloc(io_info.piece_count * sizeof(void *))))
+                HGOTO_ERROR(H5E_RESOURCE, H5E_CANTALLOC, FAIL,
+                            "memory allocation failed for read buffer list")
+        }
+
+        /* Loop with serial & single-dset read IO path */
+        for (i = 0; i < count; i++) {
+            /* Check for skipped I/O */
+            if (dset_info[i].skip_io)
+                continue;
+
+            /* set metadata tagging with dset oheader addr */
+            H5AC_tag(dset_info[i].dset->oloc.addr, &prev_tag);
+
+            /* Invoke correct "high level" I/O routine */
+            if ((*dset_info[i].io_ops.multi_read)(&io_info, &dset_info[i]) < 0)
+                HGOTO_ERROR(H5E_DATASET, H5E_READERROR, FAIL, "can't read data")
+
+            /* Reset metadata tagging */
+            H5AC_tag(prev_tag, NULL);
+        }
+
+        /* Make final multi dataset selection I/O call if we are using both
+         * features - in this case the multi_read callbacks did not perform the
+         * actual I/O */
+        H5_CHECK_OVERFLOW(io_info.piece_count, size_t, uint32_t)
+        if (!H5D_LAYOUT_CB_PERFORM_IO(&io_info))
+            if (H5F_shared_select_read(io_info.f_sh, H5FD_MEM_DRAW, (uint32_t)io_info.piece_count,
```

Comment:
Should this select read call be updated to use `pieces_added` rather than `piece_count` as well?

Reply:
Fixed, thanks